### PR TITLE
[develop] Lucero - Tests label controller

### LINKED ISSUE #33: Complete Tests Cases - Label
- [x] Functional tests
- [x] Negative tests
- [x] Smoke Tests

### PR DESCRIPTION
**Pull request by luqp**

### CODE REVIEW

File: src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java [hunk 1]
```diff
@@ -35,11 +35,29 @@ public RequestSteps(final Context context, final IRequestManager requestManager)
...
+
+    /**
+     * Sets authentication only with API key.
+     */
+    @Given("I set authentication using only a Token")
+    public void setAuthenticationToken() {
+        RequestSpecification reqSpec = RequestSpecUtils.buildOnlyToken();
+        context.setReqSpec(reqSpec);
+    }
+
```
Comment (nanrehd) [inline src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java]: I'm no really sure we should test whether there is only key or token, instead just testing if there is authorization or not, which is already covered

File: src/test/resources/features/labels/PUT_NegativeLabel.feature [hunk 2]
```diff
@@ -0,0 +1,64 @@
+@negative
+Feature: Label Controller
+
+  Background: authenticate user and set up a label
+    Given I set authentication using API key and token
+
+  @smoke
+  Scenario Outline: No gets a label with non-existent ids
```
Comment (nanrehd) [inline src/test/resources/features/labels/PUT_NegativeLabel.feature]: The sentence seems quite not understandable to me

**Review by nanrehd — COMMENTED** [review]
Great work, specially adding the methods for get requests with parameters. Left a few comments on it

File: src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java [hunk 3]
```diff
@@ -60,6 +78,18 @@ public void sendGETRequestWithParameters(final String endpoint) {
...
     }
 
+    /**
+     * Sends GET request with parameters.
+     *
+     * @param endpoint resource endpoint.
+     * @param params   request parameters.
+     */
+    @When("I send a GET request to {string} with the following parameters")
+    public void iSendAGETRequestToWithTheFollowingParameters(final String endpoint, final Map<String, String> params) {
```
Comment (carlosmc23) [inline src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java]: I think this method name is very long, I suggest sentGetWithParamameters() or something like that

File: src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java [hunk 4]
```diff
@@ -138,4 +168,16 @@ public void sendPUTRequestWithJsonBody(final String endpoint, final String body)
...
     }
+
+    /**
+     * Sends GET request with json body.
+     *
+     * @param endpoint resource endpoint.
+     * @param body     request json body.
+     */
+    @When("I send a GET request to {string} with the following json body")
+    public void iSendAGETRequestToWithTheFollowingJsonBody(final String endpoint, final String body) {
```
Comment (carlosmc23) [inline src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java]: I think this method name is very long, I suggest sentGetWithJsonBody() or something like that

File: src/test/java/org/fundacionjala/trello/stepdefs/ValidationSteps.java [hunk 5]
```diff
@@ -56,4 +58,32 @@ public void iValidateTheResponseBodyShouldMatchWithJSONSchema(final String schem
...
+        assertEquals(actualContent, content, NO_EQUALS_ERROR_MESSAGE);
+    }
+
+    /**
+     * Validates that response contains expected json content.
+     *
+     * @param content expected data.
+     */
+    @Then("I validate the response contains the following json")
+    public void validateResponseContainsJson(final String content) {
```
Comment (carlosmc23) [inline src/test/java/org/fundacionjala/trello/stepdefs/ValidationSteps.java:78]: We talked about this method in class yesterday and I remeber that you need to use jackson

File: src/test/resources/features/labels/DELETE_Label.feature [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
     Given I set authentication using API key and token
-    # default board saved as "board"
     And I have a board created
-    # default label saved as "label"
     And I have a label created
 
-  @deleteBoard
+  @functional
   Scenario: Delete a label created
+
```
Comment (carlosmc23) [inline src/test/resources/features/labels/DELETE_Label.feature:11]: you have blank lines, I suggest you delete them

File: src/test/resources/features/labels/DELETE_NegativeLabel.feature [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  Valid id match to the following Pattern: ^[0-9a-fA-F]{32}$
+  and we are validating that non-existent ids throw a 404 status code.
+
+    When I send a DELETE request to "/labels/<Non existent>"
+    Then I validate the response has status code 404
+    And I validate the response contains the following value
+    """
+    The requested resource was not found.
+    """
+
```
Comment (carlosmc23) [inline src/test/resources/features/labels/DELETE_NegativeLabel.feature:26]: you have blank lines, I suggest you delete them

File: src/test/resources/features/labels/DELETE_NegativeLabel.feature [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    Examples: Non-existent ids
+
+      | Non existent             |
+      | 000000000000000000000000 |
+      | aaaaaaaaaaaaaaaaaaaaaaaa |
+      | 00000000000fffffffffffff |
+      | 999999999999999999999999 |
+      | ffffffffffffffffffffffff |
+      | 00000000000000000000000f |
+
```
Comment (carlosmc23) [inline src/test/resources/features/labels/DELETE_NegativeLabel.feature]: you have blank lines, I suggest you delete them

File: src/test/resources/features/labels/DELETE_NegativeLabel.feature [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  Valid id match to the following Pattern: ^[0-9a-fA-F]{32}$
+  and we are validating that invalid ids throw a 400 status code.
+
+    When I send a DELETE request to "/labels/<Invalid id>"
+    Then I validate the response has status code 400
+    And I validate the response contains the following value
+    """
+    invalid id
+    """
+
```
Comment (carlosmc23) [inline src/test/resources/features/labels/DELETE_NegativeLabel.feature:38]: you have blank lines, I suggest you delete them

File: src/test/resources/features/labels/DELETE_NegativeLabel.feature [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    Examples: Invalid ids
+
```
Comment (carlosmc23) [inline src/test/resources/features/labels/DELETE_NegativeLabel.feature]: you have blank lines, I suggest you delete them

File: src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java (hunk 1 above)
Comment (luqp) [reply src/test/java/org/fundacionjala/trello/stepdefs/RequestSteps.java]: To validate that we need both keys (KEY and TOKEN) to send request,
- We should not access to user data with a token, if we are not a registered users.

File: src/test/resources/features/labels/PUT_NegativeLabel.feature (hunk 2 above)
Comment (luqp) [reply src/test/resources/features/labels/PUT_NegativeLabel.feature]: To guarantee that the PUT only updates the data of entities that exist, and does not behave improperly (boundaries that comply with the id format restriction were used).

File: build.gradle [hunk 11]
```diff
@@ -19,6 +19,7 @@ dependencies {
     implementation group: 'org.slf4j', name: 'slf4j-simple', version: '1.7.28'
     implementation group: 'io.rest-assured', name: 'rest-assured', version: '4.3.0'
     implementation group: 'io.rest-assured', name: 'json-schema-validator', version: '4.3.0'
+    implementation 'com.googlecode.json-simple:json-simple:1.1.1'
```
Comment (nanrehd) [inline build.gradle]: I'd suggest you keep the format of the other dependencies.

Comment (luqp) [reply build.gradle]: It's done!! ;)

**Review by nanrehd — APPROVED** [review]

**Review by marcos-x86 — APPROVED** [review]

**Pull request merged**